Rewrite of the Secure Websocket explanation

stakeworld opened the pull request:
Quitte a big change of text, Inspired by my own recent explorations of [setting up a rpc server](https://stakeworld.io/rpc) I think it is more informative then the old version but let me know if you have remarks or additions.

stakeworld committed:
Rewrite of the Secure Websocket explanation

stakeworld committed:
Loose the caps and domain to .io

alfarok commented:
Thanks for the updates @stakeworld!  I will give this a review tomorrow and report back with any feedback.

alfarok commented on docs/maintain/maintain-wss.md:
```diff
@@ -1,86 +1,112 @@
+to access the underlying network and/or validator node. By default you can access your node's RPC
+server from localhost (for example to rotate keys or do other maintenance) but to access it from
+from another server or for example connect to it from on UI like
+[Polkadot-JS UI](https://polkadot.js.org/apps) you will need to enable a secure proxy for the
```
```suggestion
from another server or from an applications UI (such as [Polkadot-JS UI](https://polkadot.js.org/apps)),
you will need to enable a secure proxy for the 
```

alfarok commented on docs/maintain/maintain-wss.md:
```diff
@@ -1,86 +1,112 @@
-In this guide we'll be using Ubuntu 18.04 hosted on a \$10 DigitalOcean droplet. We'll assume you're
-using a similar OS, and that you have nginx installed (if not, run `sudo apt-get install nginx`).
+:::note Enabling remote access to your validator node should not be neccessary and can only lead to
+security problems :::
```
```suggestion
:::note Enabling remote access to your validator node should not be necessary and is not
suggested as it can often lead to security problems :::
```

alfarok commented on docs/maintain/maintain-wss.md:
```diff
@@ -1,86 +1,112 @@
+example). Create a new server on your provider of choice or locally at home. See
+[Set up a Full Node](./maintain-sync) for instructions, you can choose to install from the default
+apt repository of build from scratch, whatever you prefer. In the startup options you can choose
+different settings.
```
```suggestion
Setting up any Substrate-based node relies on a similar process.  For example, they will all by default share the same websocket connection at port 9944 on localhost.
In this example, we'll set up a Kusama sync node on a debian flavoured server (such as Ubuntu 22.04). Create a new server on your provider of choice or locally at home. See
[Set up a Full Node](./maintain-sync) for additional instructions. You can choose to install from the default
apt repository or build from scratch. The startup options  in the setup process provide a variety of 
different settings that can be modified.
```

alfarok commented on docs/maintain/maintain-wss.md:
````diff
@@ -1,86 +1,112 @@
 ```
 
-This will start the syncing process with Kusama's mainnet.
+See below for explanation about the options.
````
```suggestion
The specified flag option are outlined in greater detail below.
```

alfarok commented on docs/maintain/maintain-wss.md:
```diff
@@ -1,86 +1,112 @@
+frequently done actions can be done with a pruned node, for example see account balances, make
+transfers, setup session keys, staking, etc. An archive node has the full history (database) of the
+network and can be queried in all kind of ways, give information about transfers since the network
+started, historical balances, advanced queries about past events, etc.
```
```suggestion
### Archive Node vs Pruned Node

A pruned node only tracks more recent information about the network and not its full history. Most
frequently required actions can be completed with a pruned node, such as displaying account balances, making
transfers, setting up session keys, staking, etc. An archive node has the full history (database) of the
network and can be queried in all kind of ways, such as providing historical information regarding transfers, balance histories, and more advanced queries involving past events, etc.
```

alfarok commented on docs/maintain/maintain-wss.md:
```diff
@@ -1,86 +1,112 @@
 
-:::
+Via the node startup settings you can choose **what** to expose with **how many** connections **from
+where** through your rpc server.
```
```suggestion
The node startup settings allow you to choose **what** to expose, **how many** connections to expose and **from
where** access should be granted through the rpc server.
```

alfarok commented on docs/maintain/maintain-wss.md:
```diff
@@ -1,86 +1,112 @@
+enabled proxy. So from outside one see's the ssl enabled apache2/nginx/other proxy server, which
+redirects the request to the internal rpc node. For this you will need an SSL certificate. There are
+different possible approaches; an official certificate like letsencrypt or using a self signed
+certificate.
```
```suggestion
A non secure ws port can be converted to a secure wss port by placing it behind an SSL
enabled proxy. The SSL enabled apache2/nginx/other proxy server is not externally visible, which
redirects requests to the internal rpc node. For this you will need an SSL certificate. There are
different strategies for obtaining a cert, such as using a service like letsencrypt or self signing.
```

alfarok commented on docs/maintain/maintain-wss.md:
```diff
@@ -1,86 +1,112 @@
+certificate.
 
-### Self-signed
+### Get a SSL certificate
```
```suggestion
### Obtaining an SSL Certificate
```

alfarok commented on docs/maintain/maintain-wss.md:
```diff
@@ -1,86 +1,112 @@
 
-The second approach and one we'll follow here is generating a self-signed certificate and relying on
-the raw IP address of your node when connecting to it.
+Easiest is to get a real (free) SSL certificate following LetsEncrypt instructions
```
```suggestion
One easy way to get a free SSL certificate can be achieved by following the LetsEncrypt instructions
```

alfarok commented on docs/maintain/maintain-wss.md:
```diff
@@ -127,20 +153,32 @@ server {
+Apache2 is a little heavier then nginx but also has some more tweaking posibilities. You can run it
+in different modes, prefork, worker or event. For this manual we use
+[event](https://httpd.apache.org/docs/2.4/mod/event.html) which works well on higher load
+enviroments but other modes also work fine.
```
```suggestion
Apache2 is slightly more computationally intensive when compared to nginx but also provides more configurable options. You can run it in different modes such as prefork, worker or event. In this example, we use
[event](https://httpd.apache.org/docs/2.4/mod/event.html) which works well on higher load
environments but other modes are also useful given the requirements.
```

alfarok commented on docs/maintain/maintain-wss.md:
```diff
@@ -127,20 +153,32 @@ server {
+
+The [mod_proxy_wstunnel](https://httpd.apache.org/docs/2.4/mod/mod_proxy_wstunnel.html) provides
+_support for the tunnelling of web socket connections to a backend websockets server. The connection
+is automatically upgraded to a websocket connection_. In a ssl enabled virtualhost add:
```
```suggestion
is automatically upgraded to a websocket connection_. In a SSL enabled virtualhost add:
```

alfarok commented on docs/maintain/maintain-wss.md:
````diff
@@ -1,86 +1,112 @@
 ```
 
-## Set up Nginx server
+## Install a proxy server
````
```suggestion
## Installing a Proxy Server
```

alfarok commented on docs/maintain/maintain-wss.md:
````diff
@@ -127,20 +153,32 @@ server {
+```
 
-## Connecting to the node
+## Connect to the node
````
```suggestion
## Connecting to the Node
```

alfarok reviewed (changes requested):
@stakeworld, thanks for your additions!  I added some suggestions mainly around phrasing of a few sections.  I am also going to have a teammate look through the more technical configuration details.

stakeworld committed:
Apply suggestions from code review

Co-authored-by: Keith Alfaro <keith4895@gmail.com>

stakeworld commented:
@alfarok suggestions merged

alfarok commented:
@DrW3RK who is the best person to review the node configuration?

DrW3RK commented:
@PierreBesson Could you please review this PR or assign it to someone appropriate? Thank you!

PierreBesson commented on docs/maintain/maintain-wss.md:
```diff
@@ -1,86 +1,110 @@
-You might want to host a node on one server and then connect to it from a UI hosted on another, e.g.
-[Polkadot-JS UI](https://polkadot.js.org/apps). This will not be possible unless you set up a secure
-proxy for websocket connections. Let's see how we can set up WSS on a remote Substrate node.
+A websocket provides access to the RPC server part of a substrate node, which can be used as a way
```
> A websocket provides access to the RPC server part of a substrate node

To be precise, I would phrase it as: `The substrate node RPC server can be accessed over the websocket protocol`

PierreBesson commented on docs/maintain/maintain-wss.md:
```diff
@@ -1,86 +1,110 @@
+to access the underlying network and/or validator node. By default you can access your node's RPC
+server from localhost (for example to rotate keys or do other maintenance) but to access it from
+from another server or from an applications UI (such as [Polkadot-JS UI](https://polkadot.js.org/apps)),
+you will need to enable a secure proxy for the 
```
If think it's important to point that you don't **absolutely need** a secure proxy. The "secure proxy" is recommended for enabling access to the RPC node over an TLS connection and as such encrypting the connection between the end user and RPC server. Also many browsers such as Google Chrome will block non secure WS endpoints if they come from a different origin.

PierreBesson commented on docs/maintain/maintain-wss.md:
````diff
@@ -1,86 +1,110 @@
+Or for a kusama pruned RPC node:
+
+```config
+polkadot --chain kusama --name myrpc --state-pruning 1000 --blocks-pruning archive --ws-max-connections 100 --rpc-cors all --rpc-methods Safe --ws-port 9944
````
The example is good, however I would replace references to Kusama by Polkadot as this is the polkadot docs after all :smiley:.

stakeworld replied:
Changed to polkadot. Can you confirm that `--state-pruning 1000 --blocks-pruning archive` is correct for a pruned node or should it just be `--state-pruning 1000` without the blocks-pruning?

PierreBesson commented on docs/maintain/maintain-wss.md:
```diff
@@ -1,86 +1,110 @@
+
+### Archive Node vs Pruned Node
+
+A pruned node only tracks more recent information about the network and not its full history. Most
```
Replace `only tracks more recent information about the network`  by `only keeps a limited number of finalized blocks of the network`

PierreBesson commented on docs/maintain/maintain-wss.md:
```diff
@@ -127,20 +151,31 @@ server {
+### Apache2
 
-## Importing the Certificate
+Apache2 is slightly more computationally intensive when compared to nginx but also provides more configurable options. You can run it in different modes such as prefork, worker or event. In this example, we use
```
> Apache2 is slightly more computationally intensive when compared to nginx but also provides more configurable options

I think we should refrain from doing any judgement on Apache vs nginx but ideally provide the 2 config examples.

Here is a snipped based on our internal config. Could you kindly double check it on your side and add it.

```
server {
  listen                443 ssl;
  server_name           YOUR_DOMAIN;
# for letsencrypt
  ssl_certificate /etc/letsencrypt/live/YOUR_DOMAIN/fullchain.pem;
  ssl_certificate_key /etc/letsencrypt/live/YOUR_DOMAIN/privkey.pem;
# for self-signed
  ssl_certificate /etc/ssl/certs/selfsigned.crt;
  ssl_certificate_key /etc/ssl/private/selfsigned.key;
  location / {
    limit_req zone=zone burst=5;
    proxy_buffers 16 4k;
    proxy_buffer_size 2k;
    proxy_pass ws://localhost:9944;
    proxy_http_version 1.1;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "Upgrade";
    proxy_set_header Host $host;
   }
}
```

stakeworld replied:
I made some more changes; I find the very specific ssl instruction making it more confusing. I think that somebody who sets up a rpc node will know how to setup a ssl enabled host and the specific rpc/proxypass config is the relevant part. Let me know if you disagree and think it needs to be more explicit.

PierreBesson commented on docs/maintain/maintain-wss.md:
```diff
@@ -1,86 +1,110 @@
+transfers, setting up session keys, staking, etc. An archive node has the full history (database) of the
+network and can be queried in all kind of ways, such as providing historical information regarding transfers, balance histories, and more advanced queries involving past events, etc.
+
+An archive node requires a lot more diskspace ([overview](https://stakeworld.io/snapshot)). For an
```
> [overview](https://stakeworld.io/snapshot)

Maybe let's make clear that it is an external link. For example, using a sentence like `According to statistics provided by Stakework, on Jan 6th 2023 a polkadot prune node was 100GB.`

PierreBesson reviewed (changes requested):
Thank you very much @stakeworld for the contribution.

stakeworld committed:
Add suggestions by @PierreBesson

stakeworld commented:
@PierreBesson thanks for the review and I incorporated the changes. I added some additional questions with your comments, could you take a look at them?

stakeworld committed:
Fix some broken tip remarks and minor adjustments

stakeworld committed:
Fix broken remarks

stakeworld committed:
Fix broken remarks

stakeworld commented:
@PierreBesson , wil trying out your nginx configuration I noticed two problems:
1. `proxy_pass ws://localhost:9944` isn't accepted and should be `proxy_pass http://localhost:9944;`
2. the `limit_req zone=zone burst=5;` needs a previous `limit_req_zone` declaration, could you share yours?

stakeworld committed:
Fix proxypass schema and add rate limit

stakeworld commented:
For clarity I added separate rate limiting options for nginx and apache. Also changed ws to http.

PierreBesson commented on docs/maintain/maintain-wss.md:
```diff
@@ -1,146 +1,186 @@
 
-:::note The `--rpc-cors` mode needs to be set to all so that all external connections are allowed
+An archive node requires a lot more diskspace. Start of januari 2023 a pruned polkadot takes around
+115 GB, a polkadot archive node around 765 GB, which will inrease slowly over time. For an archive
```
Minor nitpick on wording:
> Start of januari 2023 a pruned polkadot takes around
115 GB, a polkadot archive node around 765 GB, which will inrease slowly over time

> At the start of January 2023, polkadot disk usage was at 115 GB for a pruned node and 765 GB for an archive node. This value will increase with time.

stakeworld replied:
Changed

PierreBesson reviewed (changes requested):
>     proxy_pass ws://localhost:9944 isn't accepted and should be proxy_pass http://localhost:9944;
>     the limit_req zone=zone burst=5; needs a previous limit_req_zone declaration, could you share yours?

Thank for checking the config and identifying the error. As for the rate limiting, maybe we could just remove it or put it in a different section as it makes the example more complex.

But FYI, my limit_req_zone is:
```
    limit_req_zone  "$http_x_forwarded_for" zone=zone:10m rate=2r/s;
```

stakeworld commented:
> Thank for checking the config and identifying the error. As for the rate limiting, maybe we could just remove it or put it in a different section as it makes the example more complex.

I had the same idea so I already put it in a different section :)
> 
> But FYI, my limit_req_zone is:
> 
> ```
>     limit_req_zone  "$http_x_forwarded_for" zone=zone:10m rate=2r/s;
> ```

Thanks

stakeworld committed:
Textual on archive/pruned size text and minor change of limit_req_zone
